Proposed new feature: add note metadata to generated markdown files

theorib opened the pull request:
Proposed new feature:

Option to add a metadata title tag to each note including:
- Section name and number
- Lesson name and number
- Note timestamp

Presently these are added as h4 tags not for semantics but because otherwise they look huge on most generated markdown files.

The formatting logic has been placed in a new helper file but all necessary selectors needed were kept inside the `contentScripts.js` so they can all be managed in the same place.

I also proposed an amendment to the displayed text for all the options on the UI so they feel more in line with each other.

Updated UI:
![Screenshot 2024-04-02 at 15 59 07](https://github.com/CuriousAlways/udemy-notes-downloader/assets/10909928/95476a83-53f6-4613-b0c8-d4a2136ddc05)

Generated Markdown (preview mode):
![Screenshot 2024-04-02 at 16 00 08](https://github.com/CuriousAlways/udemy-notes-downloader/assets/10909928/e1411248-f086-4131-9167-1640b1ca0f5b)

Generated Markdown (raw):
![Screenshot 2024-04-02 at 16 02 08](https://github.com/CuriousAlways/udemy-notes-downloader/assets/10909928/8e4e9484-618e-421d-ad5e-60dd0fc8d8ee)

theorib committed:
feat: add note metadata to generated Markdown file

Option to Add a metadata title tag to
each note, including:
- Section name and number
- Lesson name and number
- Note timestamp

CuriousAlways commented:
first of all sorry for delayed response and Thanks for raising PR.
This looks good, i think it would nice to have clickable timestamps (i.e. link click take you to specific time on video). I did some digging lectures have following format `<link>/<lecture_id>?start=<timestamp in seconds>`. 
`edit` and `delete` notes icons  are buttons with attribute `formAction`  with this link.  We should use these links to display timestamps.

CuriousAlways commented:
<img width="1326" alt="Screenshot 2024-06-02 at 4 26 51 PM" src="https://github.com/CuriousAlways/udemy-notes-downloader/assets/58331211/050a30f7-301f-48a7-9e20-d7aa212706e4">

theorib commented:
No worries, thanks for getting back. I'll tweak the code to include a link to jump to the timestamp and send it over.

theorib commented:
I was also thinking that it would be good idea to include at least a few other of the most popular languages out there, such as Go, TypeScript, SQL, Kotlin, PHP and Bash (not a language but people are always referencing terminal commands). I could submit that as a separate PR if you like the idea.

CuriousAlways commented:
@theorib that would be awsome

theorib commented:
> <img alt="Screenshot 2024-06-02 at 4 26 51 PM" width="1326" src="https://private-user-images.githubusercontent.com/58331211/335859513-050a30f7-301f-48a7-9e20-d7aa212706e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MzQwNzYsIm5iZiI6MTcxODczMzc3NiwicGF0aCI6Ii81ODMzMTIxMS8zMzU4NTk1MTMtMDUwYTMwZjctMzAxZi00OGE3LTllMjAtZDdhYTIxMjcwNmU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE4VDE4MDI1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxYzFhNTkzMGRkZTkzZjQ3ZjdkYTg3NWI5YTM3Y2Y1YmFjODk0NjE2ZDk4ODhlOTA4NWMxZWNkZGZhYWZlNDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.j2HHsF8w5HcHYToEHDyYIqVHADVLBI2yIk73o-sUCqU">

Hey, sorry for the delay on my part this time. 

I'm looking at the source code on a Udemy's course page and I can easily find the necessary timestamp for the note written at the end of the edit, delete or the timestamp  html ID, (`id="edit-bookmark-13"` for example).

This is the timestamp value that I need in order to compose the full URL that the notes refers to, for example:
`https://www.udemy.com/course/sqlite-for-beginners/learn/lecture/12981248?start=13#notes`

What I'm struggling to find is anywhere on the source code where there is a direct reference to the actual lecture number that the note refers to and I need that to compose the whole URL (in this link example it would be `12981248` part of the URL)

Where/how did you get to that URL that you shared on your screen shot? I'm looking everywhere and can't seem to find a pattern for how I could find it or generate it. It seems like it's always created dynamically but I don't have access to it.

I'm probably missing something but If you have any pointers it would be greatly appreciated!

theorib commented:
Hi, sorry for the delay on my part this time around :)

I managed to find the URLs in the `formAction` property out of the edit or delete buttons for each note as you pointed out but there is a very strange behaviour that I haven't been able to solve:

Whenever I create a node list of all the notes, if I dig into each note, they all have their unique individual properties like the unique titles, content, etc. On the other hand the `formAction` property for the buttons on every list all have exactly the same URL. That URL points to the lesson location for the very first note on a list of notes.

It's very strange since every other aspect of the node list is unique but the URL is the same.

Here I am console.logging the data for different notes:
![Screenshot 2024-06-18 at 21 43 40](https://github.com/CuriousAlways/udemy-notes-downloader/assets/10909928/38d20e0c-cc78-4836-8d7b-317103bf62d4)

As you can see, they all have their own unique info but the URL is always the same.  
I am at a bit of a loss. I tried to find the unique URL for each note in a lot of different places without success.

Wondering if you have any ideas?

theorib committed:
feat: add additional language highlighting options
add TypeScript, Java, C#, Go, SQL, Bash, PHP and Rust
to the list of supported languages for syntax highlighting

CuriousAlways commented:
@theorib i'm not sure will check
